### PR TITLE
Dev/proterial rod4

### PR DESCRIPTION
**Pull request by ConnorGCarr**
Add Proterial (formerly Hitachi Metals) ROD-4 mass flow controller driver

### CODE REVIEW

File: docs/api/instruments/proterial/index.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+.. module:: pymeasure.instruments.proterial
+
+#########
+Proterial
+#########
+
+This section contains specific documentation on the Proterial instruments that are implemented. If you are interested in an instrument not included, please consider :doc:`adding the instrument </dev/adding_instruments/index>`.
```
Comment (BenediktBurger) [inline docs/api/instruments/proterial/index.rst:7]: Here you could mention the old name of the manufacturer.

File: pymeasure/instruments/proterial/rod4.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2023 PyMeasure Developers
```
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:4]:
```suggestion
# Copyright (c) 2013-2024 PyMeasure Developers
```

File: pymeasure/instruments/proterial/rod4.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        """Control the MFC valve mode.
+        Valid options are `flow`, `close`, and `open`. """,
+        validator=strict_discrete_set,
+        values={'flow': 0, 'close': 1, 'open': 2},
+        map_values=True
+    )
+
+    def __init__(self, *argv, **kwargs):
+        super().__init__(*argv, **kwargs)
+        self._flow_unit = None
```
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:73]: It is good to have the init at the top of the class, to find it more easily.

File: pymeasure/instruments/proterial/rod4.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            log.info("Flow units have not been set on ROD-4 channel "
+                     f"{self.id}")
+        return self._flow_unit
+
+    @flow_unit.setter
+    def flow_unit(self, units):
+        values = {'%': 0, 'sccm': 1, 'slm': 1}
+        if isinstance(units, str) and units.casefold() in values:
+            units = units.casefold()
+            self.write("\x020{ch}SFU%d" % values[units])
```
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:90]: You could use normal f-strings:
```suggestion
            self.write(f"\x020{{ch}}SFU{value[units]}")
```
If you use double curly braces, they will be reduced by one, but not evaluated.

File: pymeasure/instruments/proterial/rod4.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            self._flow_unit = 'sccm'
+        else:
+            log.info('Units must be %, sccm, or slm.')
```
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:93]: Good to log, but the user will not notice, that something is amiss. Would it not be better to raise a ValueError?

File: pymeasure/instruments/proterial/rod4.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    """
+    ch_1 = Instrument.ChannelCreator(ROD4Channel, 1)
+    ch_2 = Instrument.ChannelCreator(ROD4Channel, 2)
+    ch_3 = Instrument.ChannelCreator(ROD4Channel, 3)
+    ch_4 = Instrument.ChannelCreator(ROD4Channel, 4)
+
+    def __init__(self, adapter, name="ROD-4 MFC Controller", **kwargs):
+        super().__init__(
+            adapter, name, read_termination='\r', write_termination='\r', **kwargs
+        )
```
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:120]: It is mandatory to specify, whether it supports SCPI or not:
```suggestion
        super().__init__(
            adapter, name, read_termination='\r', write_termination='\r',
            includeSCPI=False,
            **kwargs)
```

File: tests/instruments/proterial/test_rod4.py
Comment (BenediktBurger) [inline tests/instruments/proterial/test_rod4.py:1]: Thanks for writing tests.

File: pymeasure/instruments/proterial/rod4.py (hunk 4 above)
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:90]: Why a plain property? To read just once?

Is it too much overhead to read it every time?

We try to make the instrument the single point of truth, therefore normally we read it every time from the instrument.
Someone could, for example, change the units on the instrument display, but your code will still interpret it as the old unit.

File: pymeasure/instruments/proterial/rod4.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+    @keyboard.setter
+    def keyboard(self, status):
+        values = {'unlocked': 0, 'locked': 1}
+        if isinstance(status, str) and status.casefold() in values:
+            status = status.casefold()
+            self.write("\x0200SKO%d" % values[status])
+            self._keyboard = values[status]
+        else:
+            log.info('Status must be unlocked or locked.')
```
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:144]: See above for stored state of instrument.

**Review by BenediktBurger — CHANGES_REQUESTED** [review]
Thanks for your contribution.

A few small requests for changes.

File: pymeasure/instruments/proterial/rod4.py (hunk 4 above)
Comment (ConnorGCarr) [reply pymeasure/instruments/proterial/rod4.py:90]: Good point. Oddly enough, there's no equivalent read command for the flow_unit or keyboard status. This was my workaround, but maybe it's better to just make this an instrument.setting property instead.

Comment (ConnorGCarr) [reply pymeasure/instruments/proterial/rod4.py]: Changed this to a "setting" property, removed the need for f-string

Comment (ConnorGCarr) [reply pymeasure/instruments/proterial/rod4.py]: Changed to a setting property

File: pymeasure/instruments/proterial/rod4.py (hunk 7 above)
Comment (ConnorGCarr) [reply pymeasure/instruments/proterial/rod4.py]: Changed to a setting property

File: pymeasure/instruments/proterial/rod4.py [hunk 8]
```diff
@@ -109,36 +95,28 @@ class ROD4(Instrument):
...
-    @keyboard.setter
-    def keyboard(self, status):
-        values = {'unlocked': 0, 'locked': 1}
-        if isinstance(status, str) and status.casefold() in values:
-            status = status.casefold()
-            self.write("\x0200SKO%d" % values[status])
-            self._keyboard = values[status]
-        else:
-            log.info('Status must be unlocked or locked.')
+    keyboard = Instrument.setting(
```
Comment (BenediktBurger) [inline pymeasure/instruments/proterial/rod4.py:115]: Hey,
What about `keyboard_locked`? That would make it more easy to use: `keyboard_locked = True` without thinking about the strings and what True/False mean.

Comment (ConnorGCarr) [reply pymeasure/instruments/proterial/rod4.py:115]: Thanks for the suggestion, I've changed it.

File: pymeasure/instruments/__init__.py
Comment (BenediktBurger) [inline pymeasure/instruments/__init__.py:1]: This file causes the conflict, as we do not import all manufacturers anymore.

If you remove your added line from this file, there should be no conflict anymore.

**Comment by ConnorGCarr** [comment]
> Are you ready, that I merge this driver?

Had to add set command error checking, everything is good to go now.

**Pull request merged**